File: Backend/app/services/tax_calculator.py
"""
Tax Calculation Service
Handles GST calculations (CGST, SGST, IGST, CESS)
"""

from decimal import Decimal, ROUND_HALF_UP
from typing import Dict, List, Tuple


class TaxCalculator:
    """GST tax calculator for Indian invoices"""
    
    @staticmethod
    def calculate_line_taxes(
        quantity: Decimal,
        unit_price: Decimal,
        discount: Decimal,
        gst_rate: Decimal,
        cess_rate: Decimal,
        is_inter_state: bool
    ) -> Dict[str, Decimal]:
        """
        Calculate taxes for a single line item
        
        Args:
            quantity: Item quantity
            unit_price: Unit price before tax
            discount: Discount amount
            gst_rate: GST rate percentage (e.g., 18 for 18%)
            cess_rate: Cess rate percentage
            is_inter_state: True for IGST, False for CGST+SGST
            
        Returns:
            Dictionary with taxable_amount, cgst, sgst, igst, cess, line_total
        """
        # Calculate taxable amount
        gross_amount = (quantity * unit_price).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)
        taxable_amount = (gross_amount - discount).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)
        
        # Calculate GST
        gst_amount = (taxable_amount * gst_rate / 100).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)
        
        if is_inter_state:
            # Interstate: IGST
            cgst = Decimal('0.00')
            sgst = Decimal('0.00')
            igst = gst_amount
        else:
            # Intrastate: CGST + SGST
            cgst = (gst_amount / 2).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)
            sgst = (gst_amount / 2).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)
            igst = Decimal('0.00')
        
        # Calculate CESS
        cess = Decimal('0.00')
        if cess_rate and cess_rate > 0:
            cess = (taxable_amount * cess_rate / 100).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)
        
        # Calculate line total
        line_total = taxable_amount + cgst + sgst + igst + cess
        
        return {
            'taxable_amount': taxable_amount,
            'cgst': cgst,
            'sgst': sgst,
            'igst': igst,
            'cess': cess,
            'line_total': line_total
        }
    
    @staticmethod
    def calculate_invoice_totals(
        items: List[Dict],
        company_state: str,
        client_state: str
    ) -> Dict[str, Decimal]:
        """
        Calculate total taxes for all items in an invoice
        
        Args:
            items: List of item dictionaries with quantity, unit_price, discount, gst_rate, cess_rate
            company_state: Company's state
            client_state: Client's state
            
        Returns:
            Dictionary with taxable_amount, cgst, sgst, igst, total_cess, grand_total, tax_breakdown
        """
        is_inter_state = company_state.strip().lower() != client_state.strip().lower()
        
        total_taxable = Decimal('0.00')
        total_cgst = Decimal('0.00')
        total_sgst = Decimal('0.00')
        total_igst = Decimal('0.00')
        total_cess = Decimal('0.00')
        
        # Tax breakdown by rate
        tax_breakdown: Dict[str, Dict[str, Decimal]] = {}
        
        for item in items:
            taxes = TaxCalculator.calculate_line_taxes(
                quantity=Decimal(str(item.get('quantity', 0))),
                unit_price=Decimal(str(item.get('unit_price', 0))),
                discount=Decimal(str(item.get('discount', 0))),
                gst_rate=Decimal(str(item.get('gst_rate', 0))),
                cess_rate=Decimal(str(item.get('cess_rate', 0))),
                is_inter_state=is_inter_state
            )
            
            total_taxable += taxes['taxable_amount']
            total_cgst += taxes['cgst']
            total_sgst += taxes['sgst']
            total_igst += taxes['igst']
            total_cess += taxes['cess']
            
            # Group by GST rate for breakdown
            rate_key = str(item.get('gst_rate', 0))
            if rate_key not in tax_breakdown:
                tax_breakdown[rate_key] = {
                    'rate': Decimal(str(item.get('gst_rate', 0))),
                    'taxable_amount': Decimal('0.00'),
                    'cgst': Decimal('0.00'),
                    'sgst': Decimal('0.00'),
                    'igst': Decimal('0.00'),
                    'cess': Decimal('0.00'),
                    'total_tax': Decimal('0.00')
                }
            
            tax_breakdown[rate_key]['taxable_amount'] += taxes['taxable_amount']
            tax_breakdown[rate_key]['cgst'] += taxes['cgst']
            tax_breakdown[rate_key]['sgst'] += taxes['sgst']
            tax_breakdown[rate_key]['igst'] += taxes['igst']
            tax_breakdown[rate_key]['cess'] += taxes['cess']
            tax_breakdown[rate_key]['total_tax'] += (
                taxes['cgst'] + taxes['sgst'] + taxes['igst'] + taxes['cess']
            )
        
        grand_total = total_taxable + total_cgst + total_sgst + total_igst + total_cess
        
        return {
            'taxable_amount': total_taxable,
            'cgst': total_cgst,
            'sgst': total_sgst,
            'igst': total_igst,
            'cess': total_cess,
            'grand_total': grand_total,
            'tax_breakdown': list(tax_breakdown.values())
        }


def calculate_item_taxes(
    quantity: Decimal,
    unit_price: Decimal,
    discount: Decimal,
    gst_rate: Decimal,
    cess_rate: Decimal,
    is_inter_state: bool
) -> Tuple[Decimal, Decimal, Decimal, Decimal, Decimal]:
    """
    Convenience function to calculate taxes for a single item
    
    Returns:
        Tuple of (cgst, sgst, igst, cess, line_total)
    """
    taxes = TaxCalculator.calculate_line_taxes(
        quantity, unit_price, discount, gst_rate, cess_rate, is_inter_state
    )
    return (
        taxes['cgst'],
        taxes['sgst'],
        taxes['igst'],
        taxes['cess'],
        taxes['line_total']
    )

File: src/lib/utils/tax-calculator.ts
/**
 * Tax Calculator Utility
 * 
 * Handles GST tax calculations for Indian invoices
 * - CGST + SGST for intra-state transactions (same state)
 * - IGST for inter-state transactions (different states)
 */

import { TaxBreakdown } from '@/types';

export interface TaxCalculationInput {
  amount: number;
  gstRate: number;
  quantity?: number;
  discount?: number;
  companyStateCode: string;
  clientStateCode: string;
}

export interface TaxCalculationResult {
  subtotal: number;
  discount: number;
  taxableAmount: number;
  cgst: number;
  sgst: number;
  igst: number;
  totalTax: number;
  grandTotal: number;
  isInterState: boolean;
}

/**
 * Calculate GST tax breakdown for a single item or invoice
 */
export function calculateTax(input: TaxCalculationInput): TaxCalculationResult {
  const {
    amount,
    gstRate,
    quantity = 0,
    discount = 0,
    companyStateCode,
    clientStateCode,
  } = input;

  // Calculate subtotal
  const subtotal = amount * quantity;

  // Calculate taxable amount after discount
  const taxableAmount = subtotal - discount;

  // Determine if it's inter-state (different states) or intra-state (same state)
  const isInterState = companyStateCode !== clientStateCode;

  let cgst = 0;
  let sgst = 0;
  let igst = 0;

  if (isInterState) {
    // Inter-state: Apply IGST (full GST rate)
    igst = (taxableAmount * gstRate) / 100;
  } else {
    // Intra-state: Split into CGST and SGST (half each)
    const halfRate = gstRate / 2;
    cgst = (taxableAmount * halfRate) / 100;
    sgst = (taxableAmount * halfRate) / 100;
  }

  const totalTax = cgst + sgst + igst;
  const grandTotal = taxableAmount + totalTax;

  return {
    subtotal,
    discount,
    taxableAmount,
    cgst,
    sgst,
    igst,
    totalTax,
    grandTotal,
    isInterState,
  };
}

/**
 * Calculate tax breakdown by GST rate for invoice summary
 */
export function calculateTaxBreakdown(
  items: Array<{
    amount: number;
    quantity: number;
    gstRate: number;
    discount?: number;
  }>,
  companyStateCode: string,
  clientStateCode: string
): Array<TaxBreakdown & { rate: number; taxableAmount: number }> {
  // Group items by GST rate
  const groupedByRate = items.reduce((acc, item) => {
    const rate = item.gstRate;
    if (!acc[rate]) {
      acc[rate] = [];
    }
    acc[rate].push(item);
    return acc;
  }, {} as Record<number, typeof items>);

  // Calculate tax for each rate group
  const breakdown = Object.entries(groupedByRate).map(
    ([rate, rateItems]) => {
      const gstRate = Number(rate);
      
      // Sum all items in this rate group
      const totals = rateItems.reduce(
        (sum, item) => {
          const itemCalc = calculateTax({
            amount: item.amount,
            quantity: item.quantity,
            gstRate: item.gstRate,
            discount: item.discount || 0,
            companyStateCode,
            clientStateCode,
          });

          return {
            taxable: sum.taxable + itemCalc.taxableAmount,
            cgst: sum.cgst + itemCalc.cgst,
            sgst: sum.sgst + itemCalc.sgst,
            igst: sum.igst + itemCalc.igst,
          };
        },
        { taxable: 0, cgst: 0, sgst: 0, igst: 0 }
      );

      return {
        rate: gstRate,
        taxableAmount: totals.taxable,
        cgst: totals.cgst,
        sgst: totals.sgst,
        igst: totals.igst,
        cess: 0, // Cess support can be added later if needed
        totalTax: totals.cgst + totals.sgst + totals.igst,
      };
    }
  );

  return breakdown.sort((a, b) => a.rate - b.rate);
}

/**
 * Calculate invoice totals from line items
 */
export function calculateInvoiceTotals(
  items: Array<{
    amount: number;
    quantity: number;
    gstRate: number;
    discount?: number;
  }>,
  companyStateCode: string,
  clientStateCode: string,
  additionalDiscount: number = 0,
  shippingCharges: number = 0
) {
  // Calculate all items
  const itemsTotal = items.reduce(
    (sum, item) => {
      const itemCalc = calculateTax({
        amount: item.amount,
        quantity: item.quantity,
        gstRate: item.gstRate,
        discount: item.discount || 0,
        companyStateCode,
        clientStateCode,
      });

      return {
        subtotal: sum.subtotal + itemCalc.subtotal,
        itemDiscount: sum.itemDiscount + itemCalc.discount,
        taxableAmount: sum.taxableAmount + itemCalc.taxableAmount,
        cgst: sum.cgst + itemCalc.cgst,
        sgst: sum.sgst + itemCalc.sgst,
        igst: sum.igst + itemCalc.igst,
        totalTax: sum.totalTax + itemCalc.totalTax,
      };
    },
    {
      subtotal: 0,
      itemDiscount: 0,
      taxableAmount: 0,
      cgst: 0,
      sgst: 0,
      igst: 0,
      totalTax: 0,
    }
  );

  // Apply additional invoice-level discount
  const finalTaxableAmount = itemsTotal.taxableAmount - additionalDiscount;

  // Recalculate tax if there's an additional discount
  let finalCgst = itemsTotal.cgst;
  let finalSgst = itemsTotal.sgst;
  let finalIgst = itemsTotal.igst;

  if (additionalDiscount > 0) {
    const discountRatio = finalTaxableAmount / itemsTotal.taxableAmount;
    finalCgst = itemsTotal.cgst * discountRatio;
    finalSgst = itemsTotal.sgst * discountRatio;
    finalIgst = itemsTotal.igst * discountRatio;
  }

  const finalTotalTax = finalCgst + finalSgst + finalIgst;
  const grandTotal = finalTaxableAmount + finalTotalTax + shippingCharges;

  const isInterState = companyStateCode !== clientStateCode;

  return {
    subtotal: itemsTotal.subtotal,
    itemDiscount: itemsTotal.itemDiscount,
    additionalDiscount,
    totalDiscount: itemsTotal.itemDiscount + additionalDiscount,
    taxableAmount: finalTaxableAmount,
    cgst: finalCgst,
    sgst: finalSgst,
    igst: finalIgst,
    totalTax: finalTotalTax,
    shippingCharges,
    grandTotal,
    isInterState,
  };
}

/**
 * Round to 2 decimal places (for currency)
 */
export function roundToTwo(num: number): number {
  return Math.round((num + Number.EPSILON) * 100) / 100;
}

/**
 * Calculate reverse GST (extract tax from total amount)
 * Useful when you have final price and need to break down the tax
 */
export function reverseGST(totalAmount: number, gstRate: number, isInterState: boolean) {
  const divisor = 1 + gstRate / 100;
  const baseAmount = totalAmount / divisor;
  const taxAmount = totalAmount - baseAmount;

  if (isInterState) {
    return {
      baseAmount,
      cgst: 0,
      sgst: 0,
      igst: taxAmount,
      totalTax: taxAmount,
    };
  } else {
    const halfTax = taxAmount / 2;
    return {
      baseAmount,
      cgst: halfTax,
      sgst: halfTax,
      igst: 0,
      totalTax: taxAmount,
    };
  }
}
